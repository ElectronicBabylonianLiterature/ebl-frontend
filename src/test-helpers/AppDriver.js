import React from 'react'
import { fireEvent, render, waitForElement } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import App from 'App'
import Auth from 'auth/Auth'
import WordRepository from 'dictionary/WordRepository'
import FragmentRepository from 'fragmentarium/FragmentRepository'
import ImageRepository from 'fragmentarium/ImageRepository'
import FragmentService from 'fragmentarium/FragmentService'
import WordService from 'dictionary/WordService'
import TextService from 'corpus/TextService'
import Session from 'auth/Session'
import BibliographyRepository from 'bibliography/BibliographyRepository'
import BibliographyService from 'bibliography/BibliographyService'
import { defaultErrorReporter } from 'ErrorReporterContext'
import createAuth0Config from 'auth/createAuth0Config'

function createApp(api, sessionStore) {
  const auth0Config = createAuth0Config()
  const auth = new Auth(sessionStore, defaultErrorReporter, auth0Config)
  const wordRepository = new WordRepository(api)
  const fragmentRepository = new FragmentRepository(api)
  const imageRepository = new ImageRepository(api)
  const bibliographyRepository = new BibliographyRepository(api)
  const wordService = new WordService(wordRepository)
  const bibliographyService = new BibliographyService(bibliographyRepository)
  const fragmentService = new FragmentService(
    fragmentRepository,
    imageRepository,
    bibliographyService
  )
  const textService = new TextService(api)

  return (
    <App
      auth={auth}
      wordService={wordService}
      fragmentService={fragmentService}
      bibliographyService={bibliographyService}
      textService={textService}
    />
  )
}

export default class AppDriver {
  initialEntries = []
  element = null
  session = null

  constructor(api) {
    this.api = api
  }

  withPath(path) {
    this.initialEntries = [path]
    return this
  }

  withSession() {
    this.session = new Session(
      'accessToken',
      'idToken',
      Number.MAX_SAFE_INTEGER,
      ['write:texts']
    )
    return this
  }

  render() {
    const fakeSessionStore = {
      setSession: session => {
        this.session = session
      },
      clearSession: () => {
        this.session = null
      },
      getSession: () => this.session || new Session('', '', 0, [])
    }

    this.element = render(
      <MemoryRouter initialEntries={this.initialEntries}>
        {createApp(this.api, fakeSessionStore)}
      </MemoryRouter>
    )

    return this
  }

  async waitForText(text) {
    await waitForElement(() => this.element.getByText(text))
  }

  expectTextContent(text) {
    expect(this.element.container).toHaveTextContent(text)
  }

  expectNotInContent(text) {
    expect(this.element.queryByText(text)).toBeNull()
  }

  expectLink(text, expectedHref) {
    expect(this.element.getByText(text)).toHaveAttribute('href', expectedHref)
  }

  expectBreadcrumbs(crumbs) {
    this.expectTextContent(crumbs.join(''))
  }

  expectInputElement(label, expectedValue) {
    expect(this.element.getByLabelText(label).value).toEqual(
      String(expectedValue)
    )
  }

  changeValueByLabel(label, newValue) {
    const input = this.element.getByLabelText(label)
    fireEvent.change(input, { target: { value: newValue } })
  }

  click(text, n = 0) {
    const clickable = this.element.getAllByText(text)[n]
    fireEvent.click(clickable)
    // await wait()
  }
}
