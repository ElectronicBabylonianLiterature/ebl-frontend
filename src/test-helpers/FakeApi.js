import Promise from 'bluebird'

export default class FakeApi {
  gets = []
  client = {
    fetchJson: (path, authenticate) => {
      const expectation = this.gets.find(entry => entry.path === path && entry.authenticate === authenticate)
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(new Error(`Unexpected ${authenticate ? 'authenticated' : 'not-authenticated'} fetchJson: ${path}`))
    },

    fetchBlob: (path, authenticate) => {
      return Promise.reject(new Error(`Unexpected ${authenticate ? 'authenticated' : 'not-authenticated'} fetchBlob: ${path}`))
    }
  }

  expectText (text) {
    this.gets.push({
      path: `/texts/${text.category}/${text.index}`,
      authenticate: true,
      response: text
    })
    return this
  }

  expectStatistics (statistics) {
    this.gets.push({
      path: `/statistics`,
      authenticate: false,
      response: statistics
    })
    return this
  }
}
