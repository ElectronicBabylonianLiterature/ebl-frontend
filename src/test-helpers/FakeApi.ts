import Promise from 'bluebird'
import Word from 'dictionary/domain/Word'

class Expectation {
  method: 'POST' | 'GET' = 'GET'
  path = ''
  authenticate = true
  response = {}
  verify = false
  called = false
  body: object | null = null
  isBlob = false

  constructor(data) {
    Object.assign(this, data)
  }
}

export default class FakeApi {
  private expectations: Expectation[] = []

  readonly client = {
    fetchJson: jest.fn().mockImplementation((path, authenticate) => {
      const expectation = this.expectations.find(
        entry =>
          entry.method === 'GET' &&
          entry.path === path &&
          entry.authenticate === authenticate &&
          !entry.isBlob
      )
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(
            new Error(
              `Unexpected ${
                authenticate ? 'authenticated' : 'not-authenticated'
              } fetchJson: ${path}`
            )
          )
    }),

    postJson: jest.fn().mockImplementation(path => {
      const expectation = this.expectations.find(
        entry => entry.method === 'POST' && entry.path === path && !entry.isBlob
      )
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(new Error(`Unexpected postJson: ${path}`))
    }),

    fetchBlob: jest.fn().mockImplementation((path, authenticate) => {
      const expectation = this.expectations.find(
        entry =>
          entry.method === 'GET' &&
          entry.path === path &&
          entry.authenticate === authenticate &&
          entry.isBlob
      )
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(
            new Error(
              `Unexpected ${
                authenticate ? 'authenticated' : 'not-authenticated'
              } fetchBlob: ${path}`
            )
          )
    })
  }

  expectTexts(texts): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/texts`,
        authenticate: false,
        response: texts,
        verify: true
      })
    )
    return this
  }

  allowText(text): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/texts/${text.category}/${text.index}`,
        authenticate: true,
        response: text
      })
    )
    return this
  }

  expectText(text): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/texts/${text.category}/${text.index}`,
        authenticate: true,
        response: text,
        verify: true
      })
    )
    return this
  }

  expectUpdateManuscripts(text, chapterIndex, manuscripts): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/texts/${text.category}/${text.index}/chapters/${chapterIndex}/manuscripts`,
        response: text,
        verify: true,
        body: manuscripts
      })
    )
    return this
  }

  expectUpdateLines(text, chapterIndex, lines): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/texts/${text.category}/${text.index}/chapters/${chapterIndex}/lines`,
        response: text,
        verify: true,
        body: lines
      })
    )
    return this
  }

  expectAnnotations(
    number: string,
    annotationDtos: readonly object[]
  ): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/fragments/${number}/annotations`,
        authenticate: true,
        response: { annotations: annotationDtos },
        verify: true
      })
    )
    return this
  }

  expectUpdateAnnotations(
    number: string,
    annotationDtos: readonly object[]
  ): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/fragments/${number}/annotations`,
        body: {
          fragmentNumber: number,
          annotations: annotationDtos
        },
        authenticate: true,
        response: { annotations: annotationDtos },
        verify: true
      })
    )
    return this
  }

  expectFragment(fragmentDto): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/fragments/${fragmentDto._id}`,
        authenticate: true,
        response: fragmentDto,
        verify: true
      })
    )
    return this
  }

  expectPhoto(number: string): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/fragments/${number}/photo`,
        authenticate: true,
        isBlob: true,
        verify: true
      })
    )
    return this
  }

  expectSearchWords(query: string, words: readonly Word[]): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/words?query=${encodeURIComponent(query)}`,
        response: words,
        authenticate: true,
        verify: true
      })
    )
    return this
  }

  expectWord(word: Word): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/words/${encodeURIComponent(word._id)}`,
        response: word,
        authenticate: true,
        verify: true
      })
    )
    return this
  }

  expectUpdateWord(word: Word): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/words/${encodeURIComponent(word._id)}`,
        body: word,
        response: word,
        authenticate: true,
        verify: true
      })
    )
    return this
  }

  allowStatistics(statistics): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/statistics`,
        authenticate: false,
        response: statistics
      })
    )
    return this
  }

  allowImage(file): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/images/${file}`,
        authenticate: false,
        isBlob: true
      })
    )
    return this
  }

  verifyExpectations(): void {
    const methods = {
      GET: (expectation: Expectation): void =>
        expect(
          expectation.isBlob ? this.client.fetchBlob : this.client.fetchJson
        ).toHaveBeenCalledWith(expectation.path, expectation.authenticate),
      POST: (expectation: Expectation): void =>
        expect(this.client.postJson).toHaveBeenCalledWith(
          expectation.path,
          expectation.body || expect.anything()
        )
    }
    this.expectations
      .filter(expectation => expectation.verify)
      .forEach(expectation => methods[expectation.method](expectation))
  }
}
