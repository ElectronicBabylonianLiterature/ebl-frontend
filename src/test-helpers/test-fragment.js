import {
  Folio,
  Fragment,
  Line,
  Measures,
  RecordEntry,
  Text
} from '../fragmentarium/fragment'
import { List, Map } from 'immutable'

export const fragmentDto = {
  _id: 'Test.Fragment',
  accession: '',
  cdliNumber: '',
  bmIdNumber: '',
  publication: 'Electronic Babylonian Literature',
  description: 'A fragment to be used when testing the eBL application',
  joins: [],
  length: { value: 3.14, note: '(complete)' },
  width: { value: 0.30282212, note: '' },
  thickness: {},
  collection: '',
  script: 'NB',
  notes: '',
  museum: '',
  signs: 'SAL/P₂ ŠIM GU GA\nP₅/SAL ŠIM\nŠIM ŠIM\nKU KA GA KU X',
  record: [
    { user: 'Laasonen', type: 'Revision', date: '2019-02-01T14:22:40.201231' },
    { user: 'Laasonen', type: 'Revision', date: '2019-02-01T14:23:25.214166' }
  ],
  folios: [{ name: 'FWG', number: 'M134' }],
  text: {
    lines: [
      {
        prefix: '10.',
        content: [
          {
            type: 'Word',
            value: 'sal/:',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          },
          {
            type: 'Word',
            value: 'š[im',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          },
          {
            type: 'LoneDeterminative',
            value: '{gu}',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false,
            partial: [false, true]
          },
          { type: 'Token', value: '[...]' },
          {
            type: 'Word',
            value: '.GA',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          }
        ],
        type: 'TextLine'
      },
      {
        prefix: '10.',
        content: [
          {
            type: 'Word',
            value: '::/sal',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          },
          {
            type: 'Word',
            value: 'ši]m',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          }
        ],
        type: 'TextLine'
      },
      {
        prefix: '10.',
        content: [
          {
            type: 'Word',
            value: 'šim',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          },
          { type: 'Token', value: '|' },
          {
            type: 'Word',
            value: 'šim',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          }
        ],
        type: 'TextLine'
      },
      {
        prefix: '10.',
        content: [
          { type: 'Token', value: '...' },
          {
            type: 'Word',
            value: '+ku',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          },
          {
            type: 'LoneDeterminative',
            value: '{KA.G[A}',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false,
            partial: [false, false]
          },
          { type: 'Token', value: '...' },
          { type: 'Token', value: '...' },
          {
            type: 'Word',
            value: '.ku',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          },
          {
            type: 'Word',
            value: 'x',
            uniqueLemma: [],
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          }
        ],
        type: 'TextLine'
      }
    ]
  },
  references: [
    { id: 'RN52', type: 'DISCUSSION', pages: '', notes: '', linesCited: [] }
  ],
  atf:
    '10. sal/: š[im {gu}[...].GA\n10. ::/sal ši]m\n10. šim | šim\n10. ...+ku {KA.G[A} ... ....ku x'
}

export const fragment = new Fragment({
  number: 'Test.Fragment',
  accession: '',
  cdliNumber: '',
  bmIdNumber: '',
  publication: 'Electronic Babylonian Literature',
  description: 'A fragment to be used when testing the eBL application',
  joins: List(),
  measures: Measures({
    length: 3.14,
    width: 0.30282212,
    thickness: null
  }),
  collection: '',
  script: 'NB',
  notes: '',
  museum: '',
  record: List.of(
    new RecordEntry({
      user: 'Laasonen',
      date: '2019-02-01T14:22:40.201231',
      type: 'Revision'
    }),
    new RecordEntry({
      user: 'Laasonen',
      date: '2019-02-01T14:23:25.214166',
      type: 'Revision'
    })
  ),
  folios: List([new Folio({ name: 'FWG', number: 'M134' })]),
  text: Text({
    lines: List.of(
      Line({
        type: 'TextLine',
        prefix: '10.',
        content: List.of(
          Map({
            type: 'Word',
            value: 'sal/:',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          }),
          Map({
            type: 'Word',
            value: 'š[im',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          }),
          Map({
            type: 'LoneDeterminative',
            value: '{gu}',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false,
            partial: List([false, true])
          }),
          Map({ type: 'Token', value: '[...]' }),
          Map({
            type: 'Word',
            value: '.GA',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          })
        )
      }),
      Line({
        type: 'TextLine',
        prefix: '10.',
        content: List.of(
          Map({
            type: 'Word',
            value: '::/sal',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          }),
          Map({
            type: 'Word',
            value: 'ši]m',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          })
        )
      }),
      Line({
        type: 'TextLine',
        prefix: '10.',
        content: List.of(
          Map({
            type: 'Word',
            value: 'šim',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          }),
          Map({ type: 'Token', value: '|' }),
          Map({
            type: 'Word',
            value: 'šim',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: true
          })
        )
      }),
      Line({
        type: 'TextLine',
        prefix: '10.',
        content: List.of(
          Map({ type: 'Token', value: '...' }),
          Map({
            type: 'Word',
            value: '+ku',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          }),
          Map({
            type: 'LoneDeterminative',
            value: '{KA.G[A}',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false,
            partial: List([false, false])
          }),
          Map({ type: 'Token', value: '...' }),
          Map({ type: 'Token', value: '...' }),
          Map({
            type: 'Word',
            value: '.ku',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          }),
          Map({
            type: 'Word',
            value: 'x',
            uniqueLemma: List(),
            normalized: false,
            language: 'AKKADIAN',
            lemmatizable: false
          })
        )
      })
    )
  }),
  references: List([
    Map({
      id: 'RN52',
      type: 'DISCUSSION',
      pages: '',
      notes: '',
      linesCited: List()
    })
  ]),
  atf:
    '10. sal/: š[im {gu}[...].GA\n10. ::/sal ši]m\n10. šim | šim\n10. ...+ku {KA.G[A} ... ....ku x'
})
