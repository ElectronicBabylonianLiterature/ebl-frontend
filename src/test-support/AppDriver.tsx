import React from 'react'
import {
  fireEvent,
  render,
  RenderResult,
  act,
  Matcher,
  within,
  waitFor,
} from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import _ from 'lodash'
import App from 'App'
import WordRepository from 'dictionary/infrastructure/WordRepository'
import FragmentRepository from 'fragmentarium/infrastructure/FragmentRepository'
import ApiImageRepository from 'fragmentarium/infrastructure/ImageRepository'
import FragmentService from 'fragmentarium/application/FragmentService'
import WordService from 'dictionary/application/WordService'
import TextService from 'corpus/application/TextService'
import MemorySession, { Session, guestSession } from 'auth/Session'
import BibliographyRepository from 'bibliography/infrastructure/BibliographyRepository'
import BibliographyService from 'bibliography/application/BibliographyService'
import FragmentSearchService from 'fragmentarium/application/FragmentSearchService'
import { Promise } from 'bluebird'
import { submitForm } from 'test-support/utils'
import { eblNameProperty, AuthenticationContext } from 'auth/Auth'
import SignRepository from 'signs/infrastructure/SignRepository'
import SignService from 'signs/application/SignService'

function createApp(api): JSX.Element {
  const wordRepository = new WordRepository(api)
  const fragmentRepository = new FragmentRepository(api)
  const imageRepository = new ApiImageRepository(api)
  const bibliographyRepository = new BibliographyRepository(api)
  const wordService = new WordService(wordRepository)
  const bibliographyService = new BibliographyService(bibliographyRepository)
  const fragmentService = new FragmentService(
    fragmentRepository,
    imageRepository,
    wordRepository,
    bibliographyService
  )
  const fragmentSearchService = new FragmentSearchService(fragmentRepository)
  const textService = new TextService(api, fragmentService, wordService)
  const signsRepository = new SignRepository(api)
  const signService = new SignService(signsRepository)
  return (
    <App
      signService={signService}
      wordService={wordService}
      fragmentService={fragmentService}
      fragmentSearchService={fragmentSearchService}
      bibliographyService={bibliographyService}
      textService={textService}
    />
  )
}

export default class AppDriver {
  private readonly api
  private initialEntries: string[] = []
  private view: RenderResult | null = null
  private session: Session | null = null

  constructor(api) {
    this.api = api
  }

  getView(): RenderResult {
    if (this.view) {
      return this.view
    } else {
      throw new Error('getElement called before render.')
    }
  }

  withPath(path: string): AppDriver {
    this.initialEntries = [path]
    return this
  }

  withSession(): AppDriver {
    this.session = new MemorySession([
      'write:texts',
      'read:fragments',
      'annotate:fragments',
      'read:words',
    ])
    return this
  }

  async render(): Promise<AppDriver> {
    await act(async () => {
      this.view = render(
        <MemoryRouter initialEntries={this.initialEntries}>
          <AuthenticationContext.Provider
            value={{
              login: _.noop,
              logout: _.noop,
              getSession: (): Session => this.session ?? guestSession,
              isAuthenticated: (): boolean => this.session !== null,
              getAccessToken(): Promise<string> {
                throw new Error('Not implemented')
              },
              getUser(): { [eblNameProperty]: string } {
                return { [eblNameProperty]: 'Test' }
              },
            }}
          >
            {createApp(this.api)}
          </AuthenticationContext.Provider>
        </MemoryRouter>
      )
    })

    return this
  }

  async waitForText(text: Matcher): Promise<void> {
    await this.getView().findByText(text)
  }

  async waitForTextToDisappear(text: Matcher): Promise<void> {
    await waitFor(() => {
      expect(this.getView().queryByText(text)).not.toBeInTheDocument()
    })
  }

  expectTextContent(text: string | RegExp): void {
    expect(this.getView().container).toHaveTextContent(text)
  }

  expectNotInContent(text: Matcher): void {
    expect(this.getView().queryByText(text)).not.toBeInTheDocument()
  }

  expectLink(text: Matcher, expectedHref: string): void {
    expect(this.getView().getByText(text)).toHaveAttribute('href', expectedHref)
  }

  expectBreadcrumbs(crumbs: readonly string[]): void {
    this.expectTextContent(crumbs.join(''))
  }

  expectBreadcrumb(crumb: string, link: string): void {
    expect(
      within(this.getView().getByLabelText('breadcrumb')).getByText(crumb)
    ).toHaveAttribute('href', link)
  }

  expectInputElement(label: Matcher, expectedValue: unknown): void {
    expect(this.getView().getByLabelText(label)).toHaveValue(
      String(expectedValue)
    )
  }

  expectChecked(label: Matcher): void {
    expect(this.getView().getByLabelText(label)).toBeChecked()
  }

  expectNotChecked(label: Matcher): void {
    expect(this.getView().getByLabelText(label)).not.toBeChecked()
  }

  async changeValueByLabel(label: Matcher, newValue: unknown): Promise<void> {
    const input = this.getView().getByLabelText(label)
    await act(async () => {
      fireEvent.change(input, { target: { value: newValue } })
    })
  }

  async submitForm(): Promise<void> {
    await submitForm(this.getView().container)
  }

  async click(text: Matcher, n = 0): Promise<void> {
    const clickable = this.getView().getAllByText(text)[n]
    await act(async () => {
      fireEvent.click(clickable)
    })
  }
}
