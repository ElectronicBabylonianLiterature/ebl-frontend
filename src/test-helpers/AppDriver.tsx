import React from 'react'
import {
  fireEvent,
  render,
  RenderResult,
  act,
  Matcher,
  within
} from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import _ from 'lodash'
import App from 'App'
import WordRepository from 'dictionary/infrastructure/WordRepository'
import FragmentRepository from 'fragmentarium/infrastructure/FragmentRepository'
import ApiImageRepository from 'fragmentarium/infrastructure/ImageRepository'
import FragmentService from 'fragmentarium/application/FragmentService'
import WordService from 'dictionary/application/WordService'
import TextService from 'corpus/TextService'
import MemorySession, { Session, guestSession } from 'auth/Session'
import BibliographyRepository from 'bibliography/infrastructure/BibliographyRepository'
import BibliographyService from 'bibliography/application/BibliographyService'
import FragmentSearchService from 'fragmentarium/application/FragmentSearchService'
import { Promise } from 'bluebird'
import { submitForm } from 'test-helpers/utils'
import { eblNameProperty, AuthenticationContext } from 'auth/Auth'

function createApp(api): JSX.Element {
  const wordRepository = new WordRepository(api)
  const fragmentRepository = new FragmentRepository(api)
  const imageRepository = new ApiImageRepository(api)
  const bibliographyRepository = new BibliographyRepository(api)
  const wordService = new WordService(wordRepository)
  const bibliographyService = new BibliographyService(bibliographyRepository)
  const fragmentService = new FragmentService(
    fragmentRepository,
    imageRepository,
    wordRepository,
    bibliographyService
  )
  const fragmentSearchService = new FragmentSearchService(fragmentRepository)
  const textService = new TextService(api)

  return (
    <App
      wordService={wordService}
      fragmentService={fragmentService}
      fragmentSearchService={fragmentSearchService}
      bibliographyService={bibliographyService}
      textService={textService}
    />
  )
}

export default class AppDriver {
  private readonly api
  private initialEntries: string[] = []
  private element: RenderResult | null = null
  private session: Session | null = null

  constructor(api) {
    this.api = api
  }

  getElement(): RenderResult {
    if (this.element) {
      return this.element
    } else {
      throw new Error('getElement called before render.')
    }
  }

  withPath(path): AppDriver {
    this.initialEntries = [path]
    return this
  }

  withSession(): AppDriver {
    this.session = new MemorySession([
      'write:texts',
      'read:fragments',
      'annotate:fragments',
      'read:words'
    ])
    return this
  }

  async render(): Promise<AppDriver> {
    await act(async () => {
      this.element = render(
        <MemoryRouter initialEntries={this.initialEntries}>
          <AuthenticationContext.Provider
            value={{
              login: _.noop,
              logout: _.noop,
              getSession: (): Session => this.session ?? guestSession,
              isAuthenticated: (): boolean => this.session !== null,
              getAccessToken(): Promise<string> {
                throw new Error('Not implemented')
              },
              getUser(): { [eblNameProperty]: string } {
                return { [eblNameProperty]: 'Test' }
              }
            }}
          >
            {createApp(this.api)}
          </AuthenticationContext.Provider>
        </MemoryRouter>
      )
    })

    return this
  }

  async waitForText(text: Matcher): Promise<void> {
    await this.getElement().findByText(text)
  }

  expectTextContent(text: string | RegExp): void {
    expect(this.getElement().container).toHaveTextContent(text)
  }

  expectNotInContent(text: Matcher): void {
    expect(this.getElement().queryByText(text)).toBeNull()
  }

  expectLink(text: Matcher, expectedHref: string): void {
    expect(this.getElement().getByText(text)).toHaveAttribute(
      'href',
      expectedHref
    )
  }

  expectBreadcrumbs(crumbs: readonly string[]): void {
    this.expectTextContent(crumbs.join(''))
  }

  expectBreadcrumb(crumb: string, link: string): void {
    expect(
      within(this.getElement().getByLabelText('breadcrumb')).getByText(crumb)
    ).toHaveAttribute('href', link)
  }

  expectInputElement(label: Matcher, expectedValue: unknown): void {
    expect(
      (this.getElement().getByLabelText(label) as HTMLInputElement).value
    ).toEqual(String(expectedValue))
  }

  changeValueByLabel(label: Matcher, newValue): void {
    const input = this.getElement().getByLabelText(label)
    act(() => {
      fireEvent.change(input, { target: { value: newValue } })
    })
  }

  async submitForm(): Promise<void> {
    await submitForm(this.getElement())
  }

  async click(text: Matcher, n = 0): Promise<void> {
    const clickable = this.getElement().getAllByText(text)[n]
    await act(async () => {
      fireEvent.click(clickable)
    })
  }
}
