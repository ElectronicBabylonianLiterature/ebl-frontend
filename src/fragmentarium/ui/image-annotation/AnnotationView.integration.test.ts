import AppDriver from 'test-helpers/AppDriver'
import FakeApi from 'test-helpers/FakeApi'
import { fragmentDto } from 'test-helpers/test-fragment'
import { annotationsDto } from 'test-helpers/test-annotation'
import produce from 'immer'

const fragmentWithoutReferences = produce(fragmentDto, draft => {
  draft.references = []
})
const fragmentNumber = fragmentWithoutReferences._id

let fakeApi: FakeApi
let appDriver: AppDriver

afterEach(() => {
  fakeApi.verifyExpectations()
})

describe('Diplay annotate view', () => {
  beforeEach(async () => {
    fakeApi = new FakeApi()
      .expectFragment(fragmentWithoutReferences)
      .expectPhoto(fragmentNumber)
      .expectAnnotations(fragmentNumber, annotationsDto)
    appDriver = new AppDriver(fakeApi.client)
      .withSession()
      .withPath(`/fragmentarium/${fragmentNumber}/annotate`)
      .render()
    await appDriver.waitForText(`Save`)
  })

  test('Breadcrumbs', () => {
    appDriver.expectBreadcrumbs([
      'eBL',
      'Fragmentarium',
      fragmentDto._id,
      'Annotate'
    ])
  })

  test('Snapshot', () => {
    expect(appDriver.getElement().container).toMatchSnapshot()
  })

  test('Save', () => {
    fakeApi.expectUpdateAnnotations(fragmentNumber, annotationsDto)
    appDriver.click('Save')
  })
})
