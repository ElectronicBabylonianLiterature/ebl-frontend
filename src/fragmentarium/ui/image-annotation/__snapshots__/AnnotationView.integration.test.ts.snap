// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diplay annotate view Snapshot 1`] = `
<div>
  <header
    class="Header"
  >
    <nav
      class="navbar navbar-expand-md navbar-light"
    >
      <div
        class="container"
      >
        <a
          class="active navbar-brand"
          href="/"
          title="electronic Babylonian Literature (eBL)"
        >
          <h1
            class="Header__title"
          >
            <span
              class="Header__title-main"
            >
              electronic
              <br />
              Babylonian
              <br />
              Literature
            </span>
            <small
              class="Header__title-abbreviation"
            >
              eBL
            </small>
          </h1>
        </a>
        <div
          class="navbar-collapse collapse"
          id="Header-2"
        >
          <div
            class="mx-auto navbar-nav"
          >
            <div
              class="nav-item"
            >
              <a
                aria-selected="false"
                class="nav-link"
                data-rb-event-key="/dictionary"
                href="/dictionary"
                role="tab"
                tabindex="-1"
              >
                Dictionary
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                aria-selected="false"
                class="nav-link"
                data-rb-event-key="/corpus"
                href="/corpus"
                role="tab"
                tabindex="-1"
              >
                Corpus
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                aria-selected="false"
                class="active nav-link"
                data-rb-event-key="/fragmentarium"
                href="/fragmentarium"
                role="tab"
                tabindex="-1"
              >
                Fragmentarium
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                aria-selected="false"
                class="nav-link"
                data-rb-event-key="/bibliography"
                href="/bibliography"
                role="tab"
                tabindex="-1"
              >
                Bibliography
              </a>
            </div>
          </div>
          <span
            class="navbar-text"
          >
            <button
              class="btn btn-outline-secondary btn-sm"
              type="button"
            >
              Logout
            </button>
          </span>
        </div>
        <button
          aria-controls="Header-2"
          aria-label="Toggle navigation"
          class="navbar-toggler"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
      </div>
    </nav>
  </header>
  <section
    class="App-content App-content--wide"
  >
    <header
      class="App-content__header"
    >
      <nav
        aria-label="breadcrumb"
      >
        <ol
          class="breadcrumb"
        >
          <li
            class="breadcrumb-item active"
          >
            <a
              href="/"
            >
              eBL
            </a>
          </li>
          <li
            class="breadcrumb-item active"
          >
            <a
              href="/fragmentarium"
            >
              Fragmentarium
            </a>
          </li>
          <li
            class="breadcrumb-item"
          >
            <a
              href="#"
              role="button"
            >
              Test.Fragment
            </a>
          </li>
          <li
            aria-current="page"
            class="breadcrumb-item active"
          >
            <span
              class="active"
            >
              Annotate
            </span>
          </li>
        </ol>
      </nav>
      <div
        class="float-right btn-group"
        role="group"
      />
      <h2>
        Annotate Test.Fragment
      </h2>
    </header>
    <div
      class="sc-iwsKbI bNCCdn"
    >
      <img
        alt="Test.Fragment"
        class="sc-gZMcBi gVmiLs"
        draggable="false"
      />
      <div
        class="sc-gqjmRU inYJUn"
      >
        <div
          class="sc-bZQynM yvPWU"
          style="position: absolute; left: 100%; top: 45.7%; height: 4%; width: 0.02%;"
        />
      </div>
      <div
        class="sc-gqjmRU inYJUn"
      />
      <div
        class="sc-dnqmqq ixSjBM"
      >
        Click and Drag to Annotate
      </div>
      <div />
    </div>
    <button
      class="btn btn-primary"
      type="button"
    >
      Save
    </button>
  </section>
</div>
`;
